packages/foam-core/src/model/tags.ts

```diff
@@ -0,0 +1,159 @@
+import { FoamWorkspace } from './workspace';
+import { URI } from './uri';
+import { IDisposable } from '../index';
+import { Resource } from './note';
+
+export type TagMetadata = { uri: URI };
+
+export class FoamTags implements IDisposable {
+  public readonly tags: { [key: string]: TagMetadata[] } = {};
+
+  constructor(private readonly workspace: FoamWorkspace) {}
+
+  /**
+   * List of disposables to destroy with the tags
+   */
+  private disposables: IDisposable[] = [];
+
+  /**
+   * Computes all tags in the workspace and keep them up-to-date
+   *
+   * @param workspace the target workspace
+   * @param keepMonitoring whether to recompute the links when the workspace changes
+   * @returns the FoamTags
+   */
+  public static fromWorkspace(
+    workspace: FoamWorkspace,
+    keepMonitoring: boolean = false
+  ): FoamTags {
+    let tags = new FoamTags(workspace);
+
+    Object.values(workspace.list()).forEach(resource =>
+      tags.resolveResource(resource)
+    );
+
+    if (keepMonitoring) {
+      tags.disposables.push(
+        workspace.onDidAdd(resource => {
+          tags.resolveResource(resource);
+        }),
+        workspace.onDidUpdate(change => {
+          tags.updateTagsForUpdatedResource(change.old, change.new);
+        }),
+        workspace.onDidDelete(resource => {
+          tags.updateTagsForRemovedResource(resource);
+        })
+      );
+    }
+    return tags;
+  }
+
+  updateTagsForUpdatedResource(oldVersion: Resource, newVersion: Resource) {
+    if (oldVersion.tags.size === 0 && newVersion.tags.size === 0) {
+      return;
+    }
+
+    this.resolveUpdatedTags(newVersion, oldVersion);
```

Comment:
I imagine you followed this pattern because it was the one used in `FoamGraph`.
The reason why I originally wrote it this way there was to allow for more granular updates in the graph listeners (which was never actually implemented), at the cost of the added complexity of this approach.

I wonder if for tags it would be simpler to just remove the old resource from the map, and add the new one?
This should make the code much easier to follow, wdyt?

Reply:
I am not confident if I am reading this right, so let's see if it matches. If I read it correctly, you'd propose the following 

```
loop through all tags of the old resource
for each tag, remove each instance of old resoure uri in the metadata

loop through all tags of the new resource
for each tag, add each instance of new resoure uri in the metadata
```

That would reduce a bit of the complexity of computing the difference of tags between old and new resource. Other than that, it's relatively the same. I only don't know what the impact on performance could be. But before I make a change, have I read it correctly or do you mean something else?

Reply:
yes, basically we would have a function to add a resource (a)
```
loop through all tags of the new resource
for each tag, add each instance of new resoure uri in the metadata
```

and a function to remove a resource (b)
```
loop through all tags of the old resource
for each tag, remove each instance of old resoure uri in the metadata
```

updating a resource is basically b+a

```
loop through all tags of the old resource
for each tag, remove each instance of old resoure uri in the metadata

loop through all tags of the new resource
for each tag, add each instance of new resoure uri in the metadata
```

I think we are saying the same thing, right?

Reply:
Altered! Simplified the code a bit ;-)